lib/host/src/keystone.cpp

```diff
@@ -4,111 +4,275 @@
     return KEYSTONE_ERROR;
   }
 
-  if(!runtimeFile->isValid())
-  {
+  if (!runtimeFile->isValid()) {
     ERROR("runtime file is not valid");
     return KEYSTONE_ERROR;
   }
-  if(!enclaveFile->isValid())
-  {
+  if (!enclaveFile->isValid()) {
     ERROR("enclave file is not valid");
     return KEYSTONE_ERROR;
   }
 
   /* Call Keystone Driver */
   struct keystone_ioctl_create_enclave enclp;
 
-  enclp.user_elf_ptr = (unsigned long) enclaveFile->getPtr();
-  enclp.user_elf_size = (unsigned long) enclaveFile->getSize();
-  enclp.runtime_elf_ptr = (unsigned long) runtimeFile->getPtr();
-  enclp.runtime_elf_size = (unsigned long) runtimeFile->getSize();
-
-  enclp.user_stack_size = (unsigned long) params.getEnclaveStack();
-  enclp.runtime_stack_size = (unsigned long) params.getRuntimeStack();
-
-  enclp.params.runtime_entry = (unsigned long) params.getRuntimeEntry();
-  enclp.params.user_entry = (unsigned long) params.getEnclaveEntry();
+  enclp.params.runtime_entry = (unsigned long) runtimeFile->getEntryPoint();
+  enclp.params.user_entry = (unsigned long) enclaveFile->getEntryPoint();
   enclp.params.untrusted_ptr = (unsigned long) params.getUntrustedMem();
   enclp.params.untrusted_size = (unsigned long) params.getUntrustedSize();
 
-  int ret = ioctl(fd, KEYSTONE_IOC_CREATE_ENCLAVE, &enclp);
+  enclp.min_pages = calculate_required_pages(enclaveFile->getTotalMemorySize(), params.getEnclaveStack(),
+      runtimeFile->getTotalMemorySize(), params.getRuntimeStack());
 
-  /* Files were consumed by driver and copied into epm, no longer needed */
-  delete enclaveFile;
-  delete runtimeFile;
-  runtimeFile = NULL;
-  enclaveFile = NULL;
+  runtime_stk_sz = params.getRuntimeStack();
+  enclave_stk_sz = params.getEnclaveStack();
+  untrusted_size = params.getUntrustedSize();
+  untrusted_start = params.getUntrustedMem();
+
+  /* Pass in pages to map to enclave here. */
+
+  int ret = ioctl(fd, KEYSTONE_IOC_CREATE_ENCLAVE, &enclp);
 
-  if(ret) {
+  if (ret) {
     ERROR("failed to create enclave - ioctl() failed: %d", ret);
     return KEYSTONE_ERROR;
   }
+
   eid = enclp.eid;
 
+  if(loadELF(runtimeFile) != KEYSTONE_SUCCESS) {
+    ERROR("failed to load runtime ELF");
+    return KEYSTONE_ERROR;
+  }
+
+  if(loadELF(enclaveFile) != KEYSTONE_SUCCESS) {
+    ERROR("failed to load enclave ELF");
+    return KEYSTONE_ERROR;
+  }
+
+  /* initialize stack.
+   * this will be deprecated with freemem support */
+  initStack(-1UL, runtime_stk_sz, 1);
+  initStack(enclave_stk_start, enclave_stk_sz, 0);
+
+  ret = ioctl(fd, KEYSTONE_IOC_UTM_INIT, &enclp);
+
+  if (ret) {
+    ERROR("failed to init untrusted memory - ioctl() failed: %d", ret);
+    return KEYSTONE_ERROR;
+  }
+
+  loadUntrusted();
+
+  ret = ioctl(fd, KEYSTONE_IOC_FINALIZE_ENCLAVE, &enclp);
+
+  if (ret) {
+    ERROR("failed to finalize enclave - ioctl() failed: %d", ret);
+    return KEYSTONE_ERROR;
+  }
+
+  /* ELF files are no longer needed */
+  delete enclaveFile;
```

Comment:
Make sure all the error bailouts also deallocate as needed.

Reply:
Got it.